enhancements/machine-config/securing-mcs.md

```diff
@@ -0,0 +1,112 @@
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The bootstrapping process for new cluster nodes is a tricky problem, primarily due to the fact that a new nodes start from a position of almost zero knowledge. These nodes need to fetch their machine configuration and authenticate themselves with the cluster before they can access most resources and have workloads scheduled. At the same time, a malicious actor pretending to be a new node cannot be allowed to join the cluster. Otherwise, they would be able to get access to sensitive resources and could have workloads and their secrets scheduled to them. Today, we protect against this by preventing cluster workloads from accessing the Machine Config Server, the component of the Machine Config Operator which is responsible for serving Ignition configs, derived from a set of Machine Configs, to new machines. This approach presents a problem though: workloads on OpenShift are not allowed to make use of TCP ports 22623 and 22624, the ports used by the Machine Config Server.
+
+## Motivation
+
+This blanket network policy prevents traffic destined for ports 22623 and 22624, regardless of the source and destination IP addresses. This has the potential to impact customer workloads, but also internal initiatives (e.g. OpenShift on OpenShift is complicated by this restriction).
+
+### Goals
+
+- Remove the port restriction on 22623 and 22624
+- Ensure that the node bootstrap flow remains secure
+
+### Non-Goals
+
+- Allow customers to use Machine Configs to provide pre-bootstrap configuration
+
+## Proposal
+
+If it can be ensured that no one is able to access sensitive data through the Machine Config Server, the port restrictions can be safely removed. In order to achieve this, Machine Configs can be split into two groups: pre-node-approved configs and post-node-approved configs. The pre-node-approved configs are ones which do not contain sensitive information and contain just enough to configure new nodes to submit a certificate signing request (CSR) to the API server. The post-node-approved configs contain everything else, including the overwhelming majority of customer-provided Machine Configs. When a new machine is created, it fetches its Ignition config from the Machine Config Server, which only serves the pre-node-approved configs. At this point, the machine has enough information to submit a CSR, but nothing sensitive (note that for the purposes of this proposal, the CSR bootstrap token and pull secret are not considered sensitive). Once the CSR has been approved and a kubeconfig returned, the Machine Config Daemon is able to directly enumerate Machine Configs, including the post-node-approved ones, so it then applies any remaining configuration to the machine and reboots if necessary. At this point, the machine is fully configured.
```

Comment:
> At this point, the machine has enough information to submit a CSR, but nothing sensitive (note that for the purposes of this proposal, the CSR bootstrap token and pull secret are not considered sensitive). Once the CSR has been approved

Wait... we're saying that

  - the MCS cannot distinguish a real proto-node from an attacker, _but_
  - the certificate approver _can_ distinguish a CSR sent by a real proto-node that has received non-secret information from the MCS, from a CSR sent by an attacker that has received non-secret information from the MCS

That doesn't make sense; if the certificate approver can distinguish them, then the MCS ought to be able to distinguish them in the same way. Specifically, the MCS should only respond to connections from IPs that correspond to not-yet-approved `Machine` resources (or something like that). Right?

Reply:
The logic there makes sense.  If we're willing to automatically approve a CSR, that same logic should be sufficient for accepting a connection to MCS.

However, the MCS doesn't have all the same information that the CSR approver has.  It uses the contents of the CSR itself as part of the validation, so we can't copy it easily, unless the MCS request includes more info than it does today.

For reference, see the README in https://github.com/openshift/cluster-machine-approver/ and most of the important logic is in `csr_check.go`

Reply:
> However, the MCS doesn't have all the same information that the CSR approver has. It uses the contents of the CSR itself as part of the validation, so we can't copy it easily, unless the MCS request includes more info than it does today.

But that other information doesn't matter. If the MCS can see that a request is coming from an IP that isn't the IP of a legitimate "proto-node", then it knows that the requester cannot possibly construct an approvable CSR (that points to itself), and so there is no reason for the MCS to talk to it.

Reply:
Is the source IP good enough?  Let's say untrusted-pod-A on node1 is trying to reach the MCS on node2.  Won't that request be SNAT'd on node1, so node2 is going to see node1's IP address and consider it legitimate?

Reply:
> That doesn't make sense; if the certificate approver can distinguish them, then the MCS ought to be able to distinguish them in the same way.

The client for the MCS is Ignition, which has significantly less functionality than the client for the CSR endpoint, the kubelet. Eventually, the approver could be enhanced (by us or the customer) to make use of [AWS signed requests](https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html) or [GCP instance identity verification](https://cloud.google.com/compute/docs/instances/verifying-instance-identity). Keeping the MCS out of the authentication path allows the customer to decide how much security they need for their deployment. If we did everything in the MCS, they wouldn't easily be able to swap out the implementation.

Reply:
> Is the source IP good enough? Let's say untrusted-pod-A on node1 is trying to reach the MCS on node2. Won't that request be SNAT'd on node1, so node2 is going to see node1's IP address and consider it legitimate?

node1's address is not legitimate, because node1 is an _actual_ node, not a "proto-node". An actual functioning node has no reason to talk to the MCS (as I understand it); an IP should only be contacting the MCS if there is a `Machine` that claims that IP but no `Node` that claims that IP; ie, if the cluster is expecting a node to come up with that IP, but it knows that the node hasn't actually come up yet.

(Even if node1 gets accidentally nuked and you want to recreate it, then (AIUI), you would have to delete the old node1 Node/CSR first, since otherwise the new node1's new CSR would fail to match and the approver would consider it to be an imposter. So even for reinstalls, the rule that "only IPs that correspond to a Machine but not a Node should talk to the MCS" should work.)

> Eventually, the approver could be enhanced (by us or the customer) to make use of [AWS signed requests](https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html) or [GCP instance identity verification](https://cloud.google.com/compute/docs/instances/verifying-instance-identity).

Do you mean adding that to the existing approval mechanism, or removing the existing approval mechanism entirely and switching it to be based entirely on cloud APIs? If you just mean adding, then yes, this would let the CSR approver distinguish "this is actually a real node VM" from "this is someone pretending to be a new node VM". But as long as we still have `Machine` objects, and the CSR approver still requires that `Node`s come up with IPs that match their corresponding `Machine`s, then the MCS can check those IPs itself too, and refuse to talk to any IP that isn't a valid proto-node IP.

Reply:
>  An actual functioning node has no reason to talk to the MCS (as I understand it).

Correct, unless an existing node is re-provisioning for some reason (e.g. disk failure), and the Node object still exists. (Is this a case we need to handle? Will the new CSR even be signed?)

Reply:
> ...an IP should only be contacting the MCS if there is a Machine that claims that IP but no Node that claims that IP...

Folks can also provision their own instances, without going through the machine API.  In that case, they're the hook to handle CSR signing for those machines.  With the "MCS includes CSR-approver-style-verification" approach, those folks would need a hook so they could tell the MCS, yes, really serve requests to my new proto-node I manually provisioned".

Reply:
I think it's a no-brainer to include node IP validation in the MCS - that's what my original PR has in fact here: https://github.com/openshift/machine-config-operator/pull/784

But that said, we also need to care about UPI on-premise installations where the network may be exposed to machines outside the cluster.  To me, this is where even a non-rotating auth token is extremely helpful - it provides a really strong first layer of defense.